CREATE TABLE `VARIANT` (
	`RSVR_ID` BIGINT NOT NULL PRIMARY KEY,
	`REF` VARCHAR(63) NULL,
	`CADD_PHRED` FLOAT NULL,
	`PMAF` TINYINT NOT NULL
);
CREATE TABLE `GENOTYPE` (
	`RSVR_ID` BIGINT NOT NULL, 
	`SAMPLE_ID` INTEGER NOT NULL, 
	`GT` TINYINT NOT NULL
);
CREATE TABLE `CONSEQUENCE` (
	`RSVR_ID` BIGINT NOT NULL,
	`GENE_ID` INTEGER NOT NULL,
	`TX_ID` INTEGER NOT NULL,
	`CDNA` INTEGER NULL,
	`CDS` INTEGER NULL,
	`WORST_CSQ_ID` BIGINT NOT NULL,
	`ALL_CSQ_ID` BIGINT NOT NULL,
	`CANONICAL_CSQ_ID` BIGINT NOT NULL
);
CREATE TABLE `SAMPLE` (
	`ID` INTEGER NOT NULL,
	`NAME` VARCHAR(255) NOT NULL,
	`SEX` CHAR(1) NULL,
	`ANCESTRY` VARCHAR(255) NULL,
	`UNRELATED` CHAR(1) NULL,
	`FAMILY_ID` INTEGER NULL,
	`MOTHER_ID` INTEGER NULL,
	`FATHER_ID` INTEGER NULL
);
CREATE TABLE `PHENOTYPE` (`TYPE` VARCHAR(255) NOT NULL, `CODE` VARCHAR(255) NOT NULL, `NAME` VARCHAR(1023) NOT NULL);
CREATE TABLE `SAMPLE_PHENOTYPE` ( `SAMPLE_ID` INTEGER NOT NULL, `PHENOTYPE_TYPE` VARCHAR(255) NOT NULL, `PHENOTYPE_CODE` VARCHAR(255) NOT NULL, AGE REAL NULL);

CREATE TABLE `CASE_SET` (`SAMPLE_ID` INTEGER NOT NULL, `SET_ID` VARCHAR(255) NOT NULL);
CREATE INDEX `CASE_SET-SAMPLE_ID` ON `CASE_SET`(`SAMPLE_ID`);
CREATE INDEX `CASE_SET-SET_ID` ON `CASE_SET`(`SET_ID`);

CREATE TABLE `CONSEQUENCES_LIST` ( `BIT` INTEGER NOT NULL PRIMARY KEY, `NAME` VARCHAR(255));
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (0, 'transcript_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (1, 'intron_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (2, 'exon_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (3, 'five_prime_UTR_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (4, 'three_prime_UTR_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (5, 'coding_sequence_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (6, 'synonymous_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (7, 'splice_region_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (8, 'extended_intronic_splice_region_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (9, 'exonic_splice_region_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (10, 'stop_retained_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (11, 'missense_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (12, 'conservative_missense_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (13, 'non_conservative_missense_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (14, 'inframe_deletion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (15, 'conservative_inframe_deletion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (16, 'disruptive_inframe_deletion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (17, 'inframe_insertion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (18, 'conservative_inframe_insertion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (19, 'disruptive_inframe_insertion');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (20, 'start_lost');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (21, 'stop_lost');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (22, 'frameshift_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (23, 'stop_gained');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (24, 'splice_donor_variant');
INSERT INTO CONSEQUENCES_LIST (BIT, NAME) VALUES (25, 'splice_acceptor_variant');

CREATE INDEX `CONSEQUENCES_LIST-NAME` ON CONSEQUENCES_LIST(NAME);
CREATE INDEX `VARIANT-RSVR_ID` ON `VARIANT`(`RSVR_ID`);
CREATE INDEX `GENOTYPE-RSVR_ID` ON `GENOTYPE`(`RSVR_ID`);
CREATE INDEX `GENOTYPE-SAMPLE_ID` ON `GENOTYPE`(`SAMPLE_ID`);
CREATE INDEX `CONSEQUENCE-RSVR_ID` ON `CONSEQUENCE`(`RSVR_ID`);
CREATE INDEX `CONSEQUENCE-GENE_ID` ON `CONSEQUENCE`(`GENE_ID`);
CREATE INDEX `CONSEQUENCE-TX_ID` ON `CONSEQUENCE`(`TX_ID`);
CREATE INDEX `SAMPLE-ID` ON `SAMPLE`(`ID`);
CREATE INDEX `SAMPLE-FAMILY_ID` ON `SAMPLE`(`FAMILY_ID`);
CREATE INDEX `SAMPLE-NAME` ON `SAMPLE`(`NAME`);
CREATE INDEX `PHENOTYPE-TYPE-CODE` ON `PHENOTYPE`(`TYPE`,`CODE`);
CREATE INDEX `SAMPLE_PHENOTYPE-SAMPLE_ID` ON `SAMPLE_PHENOTYPE`(`SAMPLE_ID`);
CREATE INDEX `SAMPLE_PHENOTYPE-PHENOTYPE_TYPE-PHENOTYPE_CODE` ON `SAMPLE_PHENOTYPE`(`PHENOTYPE_TYPE`, `PHENOTYPE_CODE`);

CREATE TABLE `GENE` ( `ID` INTEGER, `NAME` VARCHAR(255));
CREATE TABLE `FEATURE_TYPE` ( `ID` INTEGER, `NAME` VARCHAR(255));
CREATE TABLE `FEATURE` ( `ID` INTEGER, `TYPE` INTEGER, `START` BIGINT, `END` BIGINT, `PHASE` INTEGER, `TX_ID` INTEGER, `FIRST_CODON` VARCHAR(3), `LAST_CODON` VARCHAR(3), `CDNA_END` INTEGER, `CDS_END` INTEGER, `CDS_NUM` INTEGER);
CREATE TABLE `BIOTYPE` ( `ID` INTEGER, `NAME` VARCHAR(255));
CREATE TABLE `TX` ( `ID` INTEGER, `NAME` VARCHAR(255), `BIOTYPE_ID` INTEGER, `STRAND` INTEGER, `GENE_ID` INTEGER, `CDNA_LENGTH` INTEGER, `CDS_LENGTH` INTEGER);

CREATE INDEX `FEATURE_TYPE-ID` ON FEATURE_TYPE(ID);
CREATE INDEX `FEATURE-ID` ON FEATURE(ID);
CREATE INDEX `FEATURE-START` ON FEATURE(START);
CREATE INDEX `FEATURE-TYPE` ON FEATURE(TYPE);
CREATE INDEX `FEATURE-END` ON FEATURE(END);
CREATE INDEX `FEATURE-TX_ID` ON FEATURE(TX_ID);
CREATE INDEX `TX-ID` ON TX(ID);
CREATE INDEX `TX-GENE_ID` ON TX(GENE_ID);
CREATE INDEX `BIOTYPE-ID` ON BIOTYPE(ID);
CREATE INDEX `TX-BIOTYPE` ON TX(BIOTYPE_ID);
CREATE INDEX `GENE-ID` ON GENE(ID);
CREATE INDEX `GENE-NAME` ON GENE(NAME);
